File: src/free_proxy/services/maintenance.py
from __future__ import annotations

import asyncio
import logging
from typing import Any

from free_proxy.config import Settings
from free_proxy.domain.enums import NodeStatus, ProxyPolicyMode
from free_proxy.domain.models import MaintenanceResult, ProxyNodeRead
from free_proxy.infrastructure.database.repositories import (
    ProxyNodeRepository,
    SettingsRepository,
)
from free_proxy.services.auto_switch import AutoSwitchService
from free_proxy.services.discovery import DiscoveryService
from free_proxy.services.gateway import GatewayService
from free_proxy.services.health import MonitorState
from free_proxy.services.operations import NetworkOperationCoordinator
from free_proxy.services.probe import ProbeService
from free_proxy.services.proxy_pool import ProxyPoolService

logger = logging.getLogger(__name__)


class MaintenanceService:
    def __init__(
        self,
        app_settings: Settings,
        nodes: ProxyNodeRepository,
        settings: SettingsRepository,
        discovery: DiscoveryService,
        probe: ProbeService,
        pool: ProxyPoolService,
        gateway: GatewayService,
        auto_switch: AutoSwitchService,
        coordinator: NetworkOperationCoordinator | None = None,
    ) -> None:
        self._app_settings = app_settings
        self._nodes = nodes
        self._settings = settings
        self._discovery = discovery
        self._probe = probe
        self._pool = pool
        self._gateway = gateway
        self._auto_switch = auto_switch
        self._coordinator = coordinator
        self._lock = asyncio.Lock()

    @property
    def running(self) -> bool:
        return self._lock.locked()

    async def run(self) -> MaintenanceResult:
        if self._coordinator is not None:
            async with self._coordinator.acquire("maintenance"):
                return await self._run()
        return await self._run()

    async def _run(self) -> MaintenanceResult:
        async with self._lock:
            logger.info("Starting periodic proxy node maintenance")
            await self._nodes.clear_expired_blacklist()
            discovery_result = await self._discovery.discover()
            purge_stale = getattr(self._nodes, "purge_stale_nodes", None)
            if purge_stale is not None:
                await purge_stale(self._app_settings.stale_node_grace_seconds)
            settings = await self._settings.get()
            initial_tested: set[str] = set()
            probed_count = 0

            if (
                settings.connection_enabled
                and settings.routing_mode is not ProxyPolicyMode.FIXED
                and not self._gateway.status().active_node_id
            ):
                fast_candidates = await self._candidate_nodes()
                fast_candidates = self._pool.apply_filters(
                    fast_candidates,
                    settings,
                    include_unknown_ip_type=True,
                )
                fast_candidates.sort(key=probe_priority_key)
                fast_ids = [
                    node.id
                    for node in fast_candidates[: self._app_settings.initial_connect_test_limit]
                ]
                if fast_ids:
                    initial_tested.update(fast_ids)
                    results = await self._probe.probe_many(fast_ids)
                    probed_count += len(results)
                    if any(result.available for result in results):
                        await self._auto_switch.switch()
                        if self._gateway.status().active_node_id:
                            result = await self._result(
                                discovery_result.discovered, probed_count
                            )
                            logger.info("Maintenance completed after fast connection: %s", result)
                            return result

            remaining = [
                node.id
                for node in await self._candidate_nodes(include_unavailable=True)
                if node.id not in initial_tested
                and node.id != self._gateway.status().active_node_id
            ]
            if remaining:
                results = await self._probe.probe_many(remaining)
                probed_count += len(results)

            if settings.connection_enabled and not self._gateway.status().active_node_id:
                if settings.routing_mode is ProxyPolicyMode.FIXED and settings.fixed_node_id:
                    await self._gateway.activate(settings.fixed_node_id)
                else:
                    await self._auto_switch.switch()
            result = await self._result(discovery_result.discovered, probed_count)
            logger.info("Periodic proxy node maintenance completed: %s", result)
            return result

    async def run_job(self) -> dict[str, Any]:
        return (await self.run()).model_dump(mode="json")

    async def _candidate_nodes(self, *, include_unavailable: bool = False) -> list[ProxyNodeRead]:
        try:
            nodes = await self._nodes.list_nodes(limit=1000, current_only=True)
        except TypeError:
            nodes = await self._nodes.list_nodes(limit=1000)
        allowed_statuses = {NodeStatus.DISCOVERED, NodeStatus.READY}
        if include_unavailable:
            allowed_statuses.add(NodeStatus.UNAVAILABLE)
        return [node for node in nodes if node.status in allowed_statuses]

    async def _result(self, discovered: int, probed: int) -> MaintenanceResult:
        available = await self._nodes.count_nodes(status=NodeStatus.READY)
        return MaintenanceResult(
            discovered=discovered,
            probed=probed,
            available=available,
            connected_node_id=self._gateway.status().active_node_id,
        )


class MaintenanceMonitor:
    def __init__(
        self,
        settings: Settings,
        maintenance: MaintenanceService,
        gateway: GatewayService,
    ) -> None:
        self._settings = settings
        self._maintenance = maintenance
        self._gateway = gateway
        self._task: asyncio.Task[None] | None = None
        self.state = MonitorState()

    def start(self) -> None:
        if self._task is None:
            self._task = asyncio.create_task(self._run(), name="node-maintenance-monitor")

    @property
    def running(self) -> bool:
        return self._task is not None and not self._task.done()

    async def stop(self) -> None:
        if self._task is None:
            return
        self._task.cancel()
        try:
            await self._task
        except asyncio.CancelledError:
            pass
        self._task = None

    async def _run(self) -> None:
        while True:
            success = False
            try:
                await self._maintenance.run()
                success = True
                self.state.heartbeat(success=True)
            except asyncio.CancelledError:
                raise
            except Exception as exc:
                self.state.heartbeat(success=False, error=f"{type(exc).__name__}: {exc}")
                logger.exception("Periodic proxy node maintenance failed: %s", exc)
            if not success and not self._gateway.status().active_node_id:
                delay = self._settings.disconnected_retry_seconds
            else:
                delay = self._settings.maintenance_interval_seconds
            await asyncio.sleep(delay)


def probe_priority_key(node: ProxyNodeRead) -> tuple[int, int, int, int]:
    return (
        node.source_ping_ms or 999_999,
        -node.source_score,
        -node.source_speed_bps,
        node.source_sessions,
    )

File: internal/services/maintenance.go
package services

import (
	"context"
	"log/slog"
	"sort"
	"sync"
	"time"

	"github.com/masteralanlab/free-proxy/internal/config"
	"github.com/masteralanlab/free-proxy/internal/domain"
	"github.com/masteralanlab/free-proxy/internal/store"
)

// MaintenanceService runs the periodic discover→probe→(auto-connect) cycle.
type MaintenanceService struct {
	cfg          *config.Config
	nodes        *store.NodeRepository
	settingsRepo *store.SettingsRepository
	discovery    *DiscoveryService
	probe        *ProbeService
	pool         *ProxyPoolService
	gateway      *GatewayService
	autoSwitch   *AutoSwitchService
	coordinator  *Coordinator
	mu           sync.Mutex
}

// NewMaintenanceService constructs a MaintenanceService.
func NewMaintenanceService(cfg *config.Config, nodes *store.NodeRepository, settingsRepo *store.SettingsRepository,
	discovery *DiscoveryService, probe *ProbeService, pool *ProxyPoolService, gateway *GatewayService,
	autoSwitch *AutoSwitchService, coordinator *Coordinator) *MaintenanceService {
	return &MaintenanceService{
		cfg: cfg, nodes: nodes, settingsRepo: settingsRepo, discovery: discovery, probe: probe,
		pool: pool, gateway: gateway, autoSwitch: autoSwitch, coordinator: coordinator,
	}
}

// Run performs one maintenance cycle under the operation lock.
func (m *MaintenanceService) Run(ctx context.Context) (domain.MaintenanceResult, error) {
	var res domain.MaintenanceResult
	err := m.coordinator.Run(ctx, "maintenance", false, func(ctx context.Context) error {
		var e error
		res, e = m.run(ctx)
		return e
	})
	return res, err
}

// RunJob is the JobFunc form of Run.
func (m *MaintenanceService) RunJob(ctx context.Context) (map[string]any, error) {
	res, err := m.Run(ctx)
	if err != nil {
		return nil, err
	}
	return toMap(res)
}

func (m *MaintenanceService) run(ctx context.Context) (domain.MaintenanceResult, error) {
	m.mu.Lock()
	defer m.mu.Unlock()
	slog.Info("starting periodic maintenance", "module", "maintenance")
	_ = m.nodes.ClearExpiredBlacklist(ctx)
	discovery, err := m.discovery.Discover(ctx)
	if err != nil {
		return domain.MaintenanceResult{}, err
	}
	_, _ = m.nodes.PurgeStaleNodes(ctx, m.cfg.StaleNodeGrace())

	settings, err := m.settingsRepo.Get(ctx)
	if err != nil {
		return domain.MaintenanceResult{}, err
	}
	initialTested := map[string]bool{}
	probed := 0

	if settings.ConnectionEnabled && settings.RoutingMode != domain.PolicyFixed && m.gateway.Status().ActiveNodeID == nil {
		candidates, err := m.candidateNodes(ctx, false)
		if err != nil {
			return domain.MaintenanceResult{}, err
		}
		candidates = ApplyFilters(candidates, settings, true)
		sort.SliceStable(candidates, func(i, j int) bool { return probeLess(candidates[i], candidates[j]) })
		limit := m.cfg.InitialConnectTestLimit
		if limit > len(candidates) {
			limit = len(candidates)
		}
		var fastIDs []string
		for _, n := range candidates[:limit] {
			fastIDs = append(fastIDs, n.ID)
			initialTested[n.ID] = true
		}
		if len(fastIDs) > 0 {
			results, _ := m.probe.ProbeMany(ctx, fastIDs)
			probed += len(results)
			if anyAvailable(results) {
				_, _ = m.autoSwitch.Switch(ctx)
				if m.gateway.Status().ActiveNodeID != nil {
					return m.result(ctx, discovery.Discovered, probed)
				}
			}
		}
	}

	all, err := m.candidateNodes(ctx, true)
	if err != nil {
		return domain.MaintenanceResult{}, err
	}
	activeID := ""
	if a := m.gateway.Status().ActiveNodeID; a != nil {
		activeID = *a
	}
	var remaining []string
	for _, n := range all {
		if !initialTested[n.ID] && n.ID != activeID {
			remaining = append(remaining, n.ID)
		}
	}
	if len(remaining) > 0 {
		results, _ := m.probe.ProbeMany(ctx, remaining)
		probed += len(results)
	}

	if settings.ConnectionEnabled && m.gateway.Status().ActiveNodeID == nil {
		if settings.RoutingMode == domain.PolicyFixed && settings.FixedNodeID != nil && *settings.FixedNodeID != "" {
			_, _ = m.gateway.Activate(ctx, *settings.FixedNodeID)
		} else {
			_, _ = m.autoSwitch.Switch(ctx)
		}
	}
	return m.result(ctx, discovery.Discovered, probed)
}

func (m *MaintenanceService) candidateNodes(ctx context.Context, includeUnavailable bool) ([]domain.ProxyNodeRead, error) {
	nodes, err := m.nodes.ListNodes(ctx, store.NodeFilter{CurrentOnly: true}, 1000, 0)
	if err != nil {
		return nil, err
	}
	out := nodes[:0]
	for _, n := range nodes {
		switch n.Status {
		case domain.NodeDiscovered, domain.NodeReady:
			out = append(out, n)
		case domain.NodeUnavailable:
			if includeUnavailable {
				out = append(out, n)
			}
		}
	}
	return out, nil
}

func (m *MaintenanceService) result(ctx context.Context, discovered, probed int) (domain.MaintenanceResult, error) {
	available, err := m.nodes.CountNodes(ctx, store.NodeFilter{Status: string(domain.NodeReady)})
	if err != nil {
		return domain.MaintenanceResult{}, err
	}
	res := domain.MaintenanceResult{Discovered: discovered, Probed: probed, Available: int(available)}
	if a := m.gateway.Status().ActiveNodeID; a != nil {
		res.ConnectedNodeID = a
	}
	return res, nil
}

func anyAvailable(results []domain.ProbeResult) bool {
	for _, r := range results {
		if r.Available {
			return true
		}
	}
	return false
}

func probeLess(a, b domain.ProxyNodeRead) bool {
	pa, pb := a.SourcePingMS, b.SourcePingMS
	if pa == 0 {
		pa = 999999
	}
	if pb == 0 {
		pb = 999999
	}
	if pa != pb {
		return pa < pb
	}
	if a.SourceScore != b.SourceScore {
		return a.SourceScore > b.SourceScore
	}
	if a.SourceSpeedBPS != b.SourceSpeedBPS {
		return a.SourceSpeedBPS > b.SourceSpeedBPS
	}
	return a.SourceSessions < b.SourceSessions
}

// MaintenanceMonitor runs maintenance on an interval, backing off when disconnected.
type MaintenanceMonitor struct {
	cfg         *config.Config
	maintenance *MaintenanceService
	gateway     *GatewayService
	State       MonitorState
}

// NewMaintenanceMonitor constructs a MaintenanceMonitor.
func NewMaintenanceMonitor(cfg *config.Config, maintenance *MaintenanceService, gateway *GatewayService) *MaintenanceMonitor {
	return &MaintenanceMonitor{cfg: cfg, maintenance: maintenance, gateway: gateway}
}

// Run loops until ctx is cancelled.
func (m *MaintenanceMonitor) Run(ctx context.Context) {
	for {
		success := false
		if _, err := m.maintenance.Run(ctx); err != nil {
			m.State.Heartbeat(false, err.Error())
			slog.Warn("maintenance cycle failed", "module", "maintenance", "err", err)
		} else {
			success = true
			m.State.Heartbeat(true, "")
		}
		delay := m.cfg.MaintenanceInterval()
		if !success && m.gateway.Status().ActiveNodeID == nil {
			delay = m.cfg.DisconnectedRetry()
		}
		select {
		case <-ctx.Done():
			return
		case <-time.After(delay):
		}
	}
}
